Dockerfile

```diff
@@ -1,34 +1,38 @@
-FROM ipfs/go-ipfs:release
+FROM golang:1.9-stretch AS builder
 MAINTAINER Hector Sanjuan <hector@protocol.ai>
 
-EXPOSE 9094
-EXPOSE 9095
-EXPOSE 9096
+# This build state just builds the cluster binaries
+
+ENV GOPATH     /go
+ENV SRC_PATH   $GOPATH/src/github.com/ipfs/ipfs-cluster
+
+COPY . $SRC_PATH
+WORKDIR $SRC_PATH
+RUN make install
+
+#------------------------------------------------------
+FROM ipfs/go-ipfs
+MAINTAINER Hector Sanjuan <hector@protocol.ai>
+
+# This is the container which just puts the previously
+# built binaries on the go-ipfs-container.
 
 ENV GOPATH     /go
-ENV PATH       /go/bin:$PATH
 ENV SRC_PATH   /go/src/github.com/ipfs/ipfs-cluster
 ENV IPFS_CLUSTER_PATH /data/ipfs-cluster
 
-VOLUME $IPFS_CLUSTER_PATH
+EXPOSE 9094
+EXPOSE 9095
+EXPOSE 9096
 
-COPY . $SRC_PATH
+COPY --from=builder $GOPATH/bin/ipfs-cluster-service /usr/local/bin/ipfs-cluster-service
+COPY --from=builder $GOPATH/bin/ipfs-cluster-service /usr/local/bin/ipfs-cluster-ctl
+COPY --from=builder $SRC_PATH/docker/entrypoint.sh /usr/local/bin/start-daemons.sh
 
-RUN apk add --no-cache --virtual cluster-deps make musl-dev go git \
```

Comment:
Before we worried about installing the necessary things to build and removing them in the same RUN step so that Docker did not commit a new step (images are incremental, what's written in RUN cannot be removed later), so we saved size.

Now we create a temporary image (builder) with all the stuff, and copy out the binaries to the actual final one which is a minimal thing. This way:

* We can get rid of all the code that installed/removed things
* We can call `make install` directly
* We don't need to manually add go/make etc bc the golang image on top of which we are building already has them

Reply:
Awesome this makes more sense now